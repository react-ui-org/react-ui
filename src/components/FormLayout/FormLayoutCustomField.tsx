import React, { useContext } from 'react';
import { withGlobalProps } from '../../providers/globalProps';
import { classNames } from '../../helpers/classNames/classNames';
import { transferProps } from '../../helpers/transferProps';
import { getRootSizeClassName } from '../_helpers/getRootSizeClassName';
import { getRootValidationStateClassName } from '../_helpers/getRootValidationStateClassName';
import { isChildrenEmpty } from '../_helpers/isChildrenEmpty';
import { FormLayoutContext } from './FormLayoutContext';
import type { FormLayoutCustomFieldProps } from './FormLayoutCustomField.types';
import styles from './FormLayoutCustomField.module.scss';

const renderLabel = (
  id?: string,
  label?: string,
  labelForId?: string,
) => {
  if (labelForId && label) {
    return (
      <label
        className={styles.label}
        htmlFor={labelForId}
        id={id && `${id}__label`}
      >
        {label}
      </label>
    );
  }

  if (label) {
    return (
      <div
        className={styles.label}
        id={id && `${id}__label`}
      >
        {label}
      </div>
    );
  }

  return null;
};

export const FormLayoutCustomField: React.FunctionComponent<FormLayoutCustomFieldProps> = ({
  children = null,
  fullWidth = false,
  id,
  disabled = false,
  innerFieldSize,
  label,
  labelForId,
  required = false,
  validationState,
  ...restProps
}) => {
  const context = useContext(FormLayoutContext);

  if (isChildrenEmpty(children)) {
    return null;
  }

  return (
    <div
      {...transferProps(restProps)}
      className={classNames(
        styles.root,
        fullWidth && styles.isRootFullWidth,
        context && context.layout === 'horizontal' ? styles.isRootLayoutHorizontal : styles.isRootLayoutVertical,
        disabled && styles.isRootDisabled,
        required && styles.isRootRequired,
        getRootSizeClassName(innerFieldSize, styles),
        getRootValidationStateClassName(validationState, styles),
      )}
      id={id}
    >
      {renderLabel(id, label, labelForId)}
      <div
        className={styles.field}
        id={id && `${id}__field`}
      >
        {children}
      </div>
    </div>
  );
};

export const FormLayoutCustomFieldWithGlobalProps = withGlobalProps(FormLayoutCustomField, 'FormLayoutCustomField');

export default FormLayoutCustomFieldWithGlobalProps;
