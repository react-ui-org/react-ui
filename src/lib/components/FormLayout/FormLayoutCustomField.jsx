import PropTypes from 'prop-types';
import React, { useContext } from 'react';
import { withProviderContext } from '../../provider';
import { classNames } from '../../utils/classNames';
import { getRootSizeClassName } from '../_helpers/getRootSizeClassName';
import { getRootValidationStateClassName } from '../_helpers/getRootValidationStateClassName';
import { FormLayoutContext } from './FormLayoutContext';
import styles from './FormLayoutCustomField.scss';

const renderLabel = (id, label, labelForId) => {
  if (labelForId && label) {
    return (
      <label
        htmlFor={labelForId}
        id={id && `${id}__label`}
        className={styles.label}
      >
        {label}
      </label>
    );
  }

  if (label) {
    return (
      <div
        id={id && `${id}__label`}
        className={styles.label}
      >
        {label}
      </div>
    );
  }

  return null;
};

export const FormLayoutCustomField = ({
  children,
  fullWidth,
  id,
  disabled,
  innerFieldSize,
  label,
  labelForId,
  required,
  validationState,
}) => {
  const context = useContext(FormLayoutContext);

  return (
    <div
      id={id}
      className={classNames(
        styles.root,
        fullWidth && styles.isRootFullWidth,
        context && context.layout === 'horizontal' ? styles.rootLayoutHorizontal : styles.rootLayoutVertical,
        disabled && styles.isRootDisabled,
        required && styles.isRootRequired,
        getRootSizeClassName(innerFieldSize, styles),
        getRootValidationStateClassName(validationState, styles),
      )}
    >
      {renderLabel(id, label, labelForId)}
      <div
        id={id && `${id}__field`}
        className={styles.field}
      >
        {children}
      </div>
    </div>
  );
};

FormLayoutCustomField.defaultProps = {
  children: null,
  disabled: false,
  fullWidth: false,
  id: undefined,
  innerFieldSize: null,
  label: null,
  labelForId: undefined,
  required: false,
  validationState: null,
};

FormLayoutCustomField.propTypes = {
  /**
   * Custom HTML or React component(s).
   */
  children: PropTypes.node,
  /**
   * If `true`, label will be shown as disabled.
   */
  disabled: PropTypes.bool,
  /**
   * If `true`, the field will span the full width of its parent.
   */
  fullWidth: PropTypes.bool,
  /**
   * ID of the root HTML element.
   *
   * Also serves as base for ids of nested elements:
   * * `<ID>__field`
   * * `<ID>__label`
   */
  id: PropTypes.string,
  /**
   * Size of contained form field used to properly align label.
   */
  innerFieldSize: PropTypes.oneOf(['small', 'medium', 'large']),
  /**
   * Optional label of the field.
   */
  label: PropTypes.string,
  /**
   * Optional ID of labeled field to keep accessibility features. Only available if `label` is set.
   */
  labelForId: PropTypes.string,
  /**
   * If `true`, label will be styled as required.
   */
  required: PropTypes.bool,
  /**
   * Alter the field to provide feedback based on validation result.
   */
  validationState: PropTypes.oneOf(['invalid', 'valid', 'warning']),
};

export const FormLayoutCustomFieldWithContext = withProviderContext(FormLayoutCustomField, 'FormLayoutCustomField');

export default FormLayoutCustomFieldWithContext;
