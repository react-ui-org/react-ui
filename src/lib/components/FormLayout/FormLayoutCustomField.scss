@use "../../styles/tools/form-fields/foundation";
@use "../../styles/tools/form-fields/box-field-layout";
@use "../../styles/tools/form-fields/box-field-sizes";
@use "../../styles/tools/form-fields/variants";

// Foundation
.root {
    @include box-field-layout.in-form-layout();
    @include variants.visual(custom);
}

.label {
    @include foundation.label();
}

.isRootRequired .label {
    @include foundation.label-required();
}

// States
.isRootStateInvalid {
    @include variants.validation(invalid);
}

.isRootStateValid {
    @include variants.validation(valid);
}

.isRootStateWarning {
    @include variants.validation(warning);
}

// Layouts
.rootLayoutVertical,
.rootLayoutHorizontal {
    @include box-field-layout.vertical();
}

.rootLayoutHorizontal {
    @include box-field-layout.horizontal();
}

.rootLayoutVertical .field,
.rootLayoutHorizontal .field {
    width: auto;
}

.isRootFullWidth .field {
    justify-self: stretch;
}

// Sizes
.rootSizeSmall {
    @include box-field-sizes.size(small);
}

.rootSizeMedium {
    @include box-field-sizes.size(medium);
}

.rootSizeLarge {
    @include box-field-sizes.size(large);
}
