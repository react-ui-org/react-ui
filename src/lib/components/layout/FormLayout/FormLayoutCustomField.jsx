import PropTypes from 'prop-types';
import React from 'react';
import styles from './FormLayoutCustomField.scss';

const FormLayoutCustomField = ({
  children,
  fullWidth,
  id,
  label,
  layout,
}) => (
  <div
    id={id}
    className={`
      ${styles.root}
      ${fullWidth ? styles.isRootFullWidth : ''}
      ${layout === 'vertical' ? styles.rootLayoutVertical : styles.rootLayoutHorizontal}
    `.trim()}
  >
    {label && (
      <div
        id={id && `${id}__label`}
        className={styles.label}
      >
        {label}
      </div>
    )}
    <div
      id={id && `${id}__field`}
      className={styles.field}
    >
      {children}
    </div>
  </div>
);

FormLayoutCustomField.defaultProps = {
  children: null,
  fullWidth: false,
  id: undefined,
  label: null,
  layout: 'vertical',
};

FormLayoutCustomField.propTypes = {
  children: PropTypes.node,
  fullWidth: PropTypes.bool,
  id: PropTypes.string,
  label: PropTypes.string,
  layout: PropTypes.oneOf(['horizontal', 'vertical']),
};

export default FormLayoutCustomField;
