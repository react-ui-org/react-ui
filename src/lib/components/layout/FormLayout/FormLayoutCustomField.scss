@import '../../../styles/tools/forms/layouts';

.root {
  @include form-field-in-form-layout();
}

.label {
  overflow-wrap: break-word;
}

.rootLayoutHorizontal {
  @include form-field-layout-horizontal();
}

.rootLayoutVertical {
  @include form-field-layout-vertical();
}

.rootLayoutHorizontal .field,
.rootLayoutVertical .field {
  width: auto;
}

.isRootFullWidth .field {
  justify-self: stretch;
}

.root.rootLayoutHorizontal {
  @include breakpoint-up($form-field-horizontal-breakpoint) {
    grid-template-columns: var(--rui-local-label-width-fallback) 1fr;
    grid-template-columns: subgrid;
  }
}
