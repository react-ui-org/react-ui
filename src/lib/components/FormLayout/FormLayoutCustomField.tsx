import React, { useContext } from 'react';
import { withGlobalProps } from '../../provider';
import { transferProps } from '../_helpers/transferProps';
import { classNames } from '../../utils/classNames';
import { getRootSizeClassName } from '../_helpers/getRootSizeClassName';
import { getRootValidationStateClassName } from '../_helpers/getRootValidationStateClassName';
import { isChildrenEmpty } from '../_helpers/isChildrenEmpty';
import { FormLayoutContext } from './FormLayoutContext';
import styles from './FormLayoutCustomField.scss';
import { FormLayoutCustomFieldProps } from './FormLayout.types';

const renderLabel = (id: FormLayoutCustomFieldProps['id'], label: FormLayoutCustomFieldProps['label'], labelForId: FormLayoutCustomFieldProps['labelForId']) => {
  if (labelForId && label) {
    return (
      <label
        className={styles.label}
        htmlFor={labelForId}
        id={id && `${id}__label`}
      >
        {label}
      </label>
    );
  }

  if (label) {
    return (
      <div
        className={styles.label}
        id={id && `${id}__label`}
      >
        {label}
      </div>
    );
  }

  return null;
};

export const FormLayoutCustomField: React.FunctionComponent<FormLayoutCustomFieldProps> = ({
  children,
  disabled = false,
  fullWidth = false,
  id,
  innerFieldSize,
  label,
  labelForId,
  required = false,
  validationState,
  ...restProps
}) => {
  const context = useContext(FormLayoutContext);

  if (isChildrenEmpty(children)) {
    return null;
  }

  return (
    <div
      {...transferProps(restProps)}
      className={classNames(
        styles.root,
        fullWidth && styles.isRootFullWidth,
        context && context.layout === 'horizontal' ? styles.isRootLayoutHorizontal : styles.isRootLayoutVertical,
        disabled && styles.isRootDisabled,
        required && styles.isRootRequired,
        getRootSizeClassName(innerFieldSize, styles),
        getRootValidationStateClassName(validationState, styles),
      )}
      id={id}
    >
      {renderLabel(id, label, labelForId)}
      <div
        className={styles.field}
        id={id && `${id}__field`}
      >
        {children}
      </div>
    </div>
  );
};

export const FormLayoutCustomFieldWithGlobalProps = withGlobalProps(FormLayoutCustomField, 'FormLayoutCustomField');

export default FormLayoutCustomFieldWithGlobalProps;
