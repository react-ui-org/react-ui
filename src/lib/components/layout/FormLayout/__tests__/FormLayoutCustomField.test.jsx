import React from 'react';
import { shallow } from 'enzyme';
import FormLayoutCustomField from '../FormLayoutCustomField';

describe('rendering', () => {
  it('renders correctly with a single child', () => {
    const tree = shallow((
      <FormLayoutCustomField>
        <span>Custom text in form</span>
      </FormLayoutCustomField>
    ));

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with multiple children', () => {
    const tree = shallow((
      <FormLayoutCustomField>
        <span>Custom text in form 1</span>
        <span>Custom text in form 2</span>
        <span>Custom text in form 3</span>
      </FormLayoutCustomField>
    ));

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with all props', () => {
    const tree = shallow((
      <FormLayoutCustomField
        fullWidth
        label="Label"
        id="my-custom-field"
        layout="horizontal"
      >
        <span>Custom text in form 1</span>
        <span>Custom text in form 2</span>
        <span>Custom text in form 3</span>
      </FormLayoutCustomField>
    ));

    expect(tree).toMatchSnapshot();
  });
});
