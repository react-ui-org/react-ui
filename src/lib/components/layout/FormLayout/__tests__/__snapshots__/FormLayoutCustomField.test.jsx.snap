// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders correctly with a single child 1`] = `
<div
  className="root
      
      rootLayoutVertical"
>
  <div
    className="field"
  >
    <span>
      Custom text in form
    </span>
  </div>
</div>
`;

exports[`rendering renders correctly with all props 1`] = `
<div
  className="root
      isRootFullWidth
      rootLayoutHorizontal"
  id="my-custom-field"
>
  <div
    className="label"
    id="my-custom-field__label"
  >
    Label
  </div>
  <div
    className="field"
    id="my-custom-field__field"
  >
    <span>
      Custom text in form 1
    </span>
    <span>
      Custom text in form 2
    </span>
    <span>
      Custom text in form 3
    </span>
  </div>
</div>
`;

exports[`rendering renders correctly with multiple children 1`] = `
<div
  className="root
      
      rootLayoutVertical"
>
  <div
    className="field"
  >
    <span>
      Custom text in form 1
    </span>
    <span>
      Custom text in form 2
    </span>
    <span>
      Custom text in form 3
    </span>
  </div>
</div>
`;
